add rsync client and another constructor for rsync server

alaypatel07 opened the pull request:
**Describe what this PR does**
This PR adds Rsync client implementations

**Is there anything that requires special attention?**
This depends on changes in #13 

**Related issues:**
the second half of #12

alaypatel07 committed:
add rsync client implementation

Signed-off-by: Alay Patel <alay1431@gmail.com>

alaypatel07 committed:
add NewServerWithStunnelRoute to rsync/server.go

Signed-off-by: Alay Patel <alay1431@gmail.com>

alaypatel07 committed:
convert bool to bool ptr in transfer Status

This avoid defaulting bools to a value if unused.

Signed-off-by: Alay Patel <alay1431@gmail.com>

JohnStrunk commented on transfer/transfer.go:
```diff
@@ -78,8 +78,8 @@ type Running struct {
 type Completed struct {
-	Successful bool
-	Failure    bool
+	Successful *bool
+	Failure    *bool
```
Given the change to *bool, what are the semantics? Ex: Is false, false different than nil, nil?

I think that also raises a broader question of the semantics of the Status type, in general. There are a lot of pointers in there, and I'm wondering if it could be simplified a bit:

```golang
type Status struct {
    StartedAt *metav1.Time // nil means it hasn't started, other fields nil also.
    FinishedAt *metav1.Time // nil means it hasn't finished (still running if StartedAt != nil)
    // if FinishedAt != nil, one of the following 2 would be true, else both false
    Successful bool
    Failed bool
}
```

alaypatel07 replied:
The change was an oversight. Didn't realize we had a higher level pointer for differentiating.

While the suggested semantics, works for now is it, I am hoping we could get to something like the following as we add more features to the client status. Let me know if that makes sense, I can drop the changes in this PR to convert this to *bool.

```
type Status struct {
	Running   *Running // if this is not nil, other two are
	Completed *Completed // if this is not nil, other two are
	Pending   *Pending // if this is not nil, other two are
	// if all three are nil, other errors like unable to get to apiserver may be categorize them in Unknown?
}

type Running struct {
	StartedAt          *metav1.Time
	// ProgressPercentage string 
}

type Completed struct {
	Successful   bool
	Failure      bool
	FinishedAt   *metav1.Time
	StartedAt    *metav1.Time
	//ExitCode     int
	//LastLogLines [10]string
}
//
//type Pending struct {
//	Reason string
//}
```

alaypatel07 replied:
Thinking about this more, I think it makes more sense to consolidate on what you proposed:

```Golang
type Status struct {
    StartedAt *metav1.Time // nil means it hasn't started, other fields nil also.
    FinishedAt *metav1.Time // nil means it hasn't finished (still running if StartedAt != nil)
    LastLogLines [10]string{} //will be filled as soon as StartedAt is not nil
    // if FinishedAt != nil the following fields would be meaningful
    Successful bool
    Failed bool
    ExitCode int 
}
```

alaypatel07 replied:
Created issue https://github.com/backube/pvc-transfer/issues/15 for covering future work that can be added

JohnStrunk commented on transfer/rsync/client.go:
```diff
@@ -0,0 +1,431 @@
+	}
+
+	for _, pod := range podList.Items {
+		if pod.Status.ContainerStatuses != nil || len(pod.Status.ContainerStatuses) > 0 {
```
you mean && here? If so, this can be shortened to just the length test:
https://go.dev/play/p/wuLmFqHCuVm

alaypatel07 replied:
+1 I agree, will do it. Thanks for the quick test.

JohnStrunk commented on transfer/rsync/client.go:
```diff
@@ -0,0 +1,431 @@
+	var namespace string
+	namespaces := pvcList.Namespaces()
+	if len(namespaces) > 0 {
+		namespace = pvcList.Namespaces()[0]
```
```suggestion
		namespace = namespaces[0]
```

JohnStrunk commented on transfer/rsync/client.go:
```diff
@@ -0,0 +1,431 @@
+				Env: []corev1.EnvVar{
+					{
+						Name:  "RSYNC_PASSWORD",
+						Value: tc.password,
```
Password should come in via a Secret. It'll be visible to anyone w/ Get permission on the NS currently.
https://github.com/backube/volsync/blob/5fcc3961cd3c79a62ef440d9470389fe101b373e/controllers/utils/utils.go#L55-L68

JohnStrunk commented on transfer/rsync/client.go:
```diff
@@ -0,0 +1,431 @@
+			{
+				Name: "rsync-communication",
+				VolumeSource: corev1.VolumeSource{
+					EmptyDir: &corev1.EmptyDirVolumeSource{Medium: corev1.StorageMediumDefault},
```
Assuming this is just a socket:
```suggestion
					EmptyDir: &corev1.EmptyDirVolumeSource{Medium: corev1.StorageMediumMemory},
```

JohnStrunk commented on transfer/rsync/client.go:
```diff
@@ -0,0 +1,431 @@
+					},
+					{
+						Name:      "rsync-communication",
+						MountPath: "/usr/share/rsync",
```
Assuming this needs to match L322, you may want to make it a const to be used in both places

JohnStrunk commented on transfer/rsync/client.go:
```diff
@@ -0,0 +1,431 @@
+				Resources: []string{"securitycontextconstraints"},
+				// Must match the name of the SCC that is deployed w/ the operator
+				// config/openshift/mover_scc.yaml
+				ResourceNames: []string{DefaultSCCName},
```
We're going to need to find a way to deal with this. Looks like it's a hardcoded name, but that may not work well for library users.

alaypatel07 committed:
- drop *bool in status
- refactor if condition for len of containerStatuses
- store rsync password in secret
- drop get from get* methods
- use more constants

Signed-off-by: Alay Patel <alay1431@gmail.com>

alaypatel07 committed:
transfer: add SCCName to PodOptions and implement in rsync pkg

Signed-off-by: Alay Patel <alay1431@gmail.com>

JohnStrunk commented on transfer/rsync/server.go:
```diff
@@ -254,6 +303,12 @@ func NewServer(ctx context.Context, c client.Client, logger logr.Logger,
 	}
 
+	if podOptions.SCCName == nil || (podOptions.SCCName != nil && *podOptions.SCCName == "") {
+		//TODO: raise a warning event
```
I'm not sure whether this would be a warning or not... If you're not on OpenShift, lack of an SCC name doesn't matter.

JohnStrunk reviewed (commented):
/lgtm
